File: HOT100/6.矩阵/48.旋转图像.cpp
/*
 * @lc app=leetcode.cn id=48 lang=cpp
 *
 * [48] 旋转图像
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
// @lc code=start
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for (int i = 0; i < n; i ++ )
        {
            for (int j = 0; j < i; j ++ )
            {
                swap(matrix[i][j], matrix[j][i]);
            }
        }
        for (auto& row : matrix)
        {
            ranges::reverse(row);
        }
    }
};
// @lc code=end

class Solution {
    public:
        void rotate(vector<vector<int>>& matrix) {
            int m = matrix.size() - 1, n = matrix[0].size() - 1;
            int l = 0, r = n, t = 0, b = m;
            int left = 0, right = n;
            while (left <= right && l <= r && t <= b)
            {
                for (int i = left; i < right; i ++ )
                {
                    int tmp = matrix[t][l + i];
                    matrix[t][l + i] = matrix[b - i][l];
                    matrix[b - i][l] = matrix[b][r - i];
                    matrix[b][r - i] = matrix[t + i][r];
                    matrix[t + i][r] = tmp;
                }
                l ++, r --, t ++, b --;
                right -= 2;
            }
        }
    };

File: HOT100/6.2/48.旋转图像.py
#
# @lc app=leetcode.cn id=48 lang=python3
#
# [48] 旋转图像
#
from mytools import *
# @lc code=start
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        tmp =  [list(row)[::-1] for row in zip(*matrix)]
        matrix = tmp
# @lc code=end
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(n):
            for j in range(i + 1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        for row in matrix:
            row.reverse()
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        cnt = 0
        while cnt < n:
            i, j = cnt, cnt
            for _ in range(n - cnt):
                matrix[i][cnt], matrix[cnt][j] = matrix[cnt][j], matrix[i][cnt]
                i += 1; j += 1
            cnt += 1
        l, r = 0, n - 1
        while l < r:
            for i in range(n):
                matrix[i][l], matrix[i][r] = matrix[i][r], matrix[i][l]
            l += 1
            r -= 1